<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Melanie Peña Torres
- Nombre de alumno 2: Valentina Rojas Osorio


### **Link de repositorio de GitHub:** `http://....`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [75]:
!pip install -r requirements.txt

In [76]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
#!pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [77]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [78]:
df_grades = pd.read_json('data/students_grades.json')
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **875** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **875** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier alumno que posea un valor diferente a esta escala, debería ser eliminado de la base de datos.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [79]:
from IPython.core.display import Markdown

df_grades[["math score", "reading score", "writing score"]] = df_grades[["math score", "reading score", "writing score"]].apply(pd.to_numeric)

def exploratory_data_analysis(dataframe):
    
    display(Markdown(f"##### 1. El DataFrame tiene {dataframe.shape[0]} filas y {dataframe.shape[1]} columnas"))
    
    display(Markdown(f"##### 2. Columnas del DataFrame:\n{dataframe.columns.tolist()}"))
    
    display(Markdown("##### 3. Primeras 5 filas:"))
    display(dataframe.head(5))
    
    display(Markdown("##### Últimas 5 filas:"))
    display(dataframe.tail(5))
    
    display(Markdown("##### Muestreo aleatorio de 5 filas:"))
    display(dataframe.sample(5))
    
    display(Markdown("##### 4. Descripción numérica del DataFrame:"))
    display(dataframe[['math score', 'reading score', 'writing score']].describe())
    
    display(Markdown("##### 5. Cantidad de valores nulos por columna:"))
    display(dataframe.isna().sum())
    
    display(Markdown("##### 6. Cantidad de valores únicos por columna:"))
    display(dataframe.nunique())
    
    display(Markdown(f"##### 7. Total de filas con el nombre duplicado: {dataframe.duplicated(subset=['names']).sum()}"))
    
    display(Markdown("##### 8. Filas duplicadas por el nombre:"))
    display(dataframe[dataframe.duplicated(subset=["names"])])




In [80]:
exploratory_data_analysis(df_grades)

##### 1. El DataFrame tiene 1400 filas y 9 columnas

##### 2. Columnas del DataFrame:
['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

##### 3. Primeras 5 filas:

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard                    none       124.0          142.0          142.0  
2  standard                    none       156.0          196.0          193.0  
3  standard               completed        80.0           78.0           81.0  
4  standard                    none       131.0          166.0          166.0

##### Últimas 5 filas:

names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1395      standard                    none       193.0          142.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1398  free/reduced                    none       193.0          196.0   
1399      standard               completed        68.0           80.0   

      writing score  
1395          193.0  
1396           81.0  
1397           62.0  
1398          193.0  
1399           76.0

##### Muestreo aleatorio de 5 filas:

names  gender race/ethnicity parental level of education  \
1251    Kathryn Neville  female        group C                some college   
470         Evan Ammons    male        group D           bachelor's degree   
368   Aurelia Spaulding  female        group B                 high school   
14        Barbara Clerk  female        group D            some high school   
947          Ira Mercer    male        group C            some high school   

             lunch test preparation course  math score  reading score  \
1251      standard                    none        55.0           69.0   
470   free/reduced                    none        68.0           68.0   
368   free/reduced               completed        46.0           54.0   
14        standard                    none       193.0          171.0   
947   free/reduced               completed        56.0           61.0   

      writing score  
1251           65.0  
470            67.0  
368            58.0  
14            131.0  
947            60.0

##### 4. Descripción numérica del DataFrame:

math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

##### 5. Cantidad de valores nulos por columna:

names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64

##### 6. Cantidad de valores únicos por columna:

names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64

##### 7. Total de filas con el nombre duplicado: 247

##### 8. Filas duplicadas por el nombre:

names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course  math score  reading score  \
12        standard                    none       193.0          193.0   
81        standard                    none        82.0           85.0   
160       standard               completed        60.0           70.0   
170   free/reduced                    none        43.0           60.0   
176       standard                    none        54.0           61.0   
...            ...                     ...         ...            ...   
1385      standard                    none       193.0          147.0   
1390      standard                    none       124.0           64.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1398  free/reduced                    none       193.0          196.0   

      writing score  
12            193.0  
81             87.0  
160            74.0  
170            58.0  
176            58.0  
...             ...  
1385          131.0  
1390          193.0  
1393           49.0  
1395          193.0  
1398          193.0  

[247 rows x 9 columns]

In [81]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

**Justificación:**

9. La anomalía es que existen valores 'none' pero pandas no es capaz de identificarlos cono Nan, por lo que en la fila test preparation course no nos sale ningún valor, para arreglar eso, se debería reeplazar los 'none' de los json por valores pd.NaN


10. Los promedios o medianas para math, writing y reading son relativamente similares, situándose alrededor de 75 y 76 puntos, lo que indica que el rendimiento de los estudiantes en estas tres áreas no presenta diferencias significativas. Sin embargo, al examinar los cuartiles superiores e inferiores, se observa una dispersión en los puntajes obtenidos, lo cual indica una heterogeneidad evidente en el rendimiento individual dentro de cada grupo, también se pueden identificar outliers y errores en los datos, ya que existen puntajes superiores a 100, lo cual no es posible según la escala de calificación establecida de 0 a 100, los que pueden deberse a mal tipeo de datos.

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y cree una función que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**

- [ ] Crear función que permita limpiar el dataset.
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hint**

Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.


**Respuesta Esperada:**

In [83]:
def clean_data(dataframe_in):
    dataframe = dataframe_in.copy()
    dataframe = dataframe.drop_duplicates(subset=['names'])
    dataframe = dataframe.dropna()
    dataframe = dataframe[
        (dataframe.iloc[:, 6] < 101)
        & (dataframe.iloc[:, 7] < 101)
        & (dataframe.iloc[:, 8] < 101)
    ].reset_index(drop=True)
    return dataframe

In [84]:
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        37.0           46.0   
1        standard               completed        80.0           78.0   
2        standard                    none        73.0           66.0   
3        standard               completed        81.0           75.0   
4        standard               completed        94.0           86.0   
..            ...                     ...         ...            ...   
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing score  
0             46.0  
1             81.0  
2             63.0  
3             76.0  
4             87.0  
..             ...  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0  

[875 rows x 9 columns]

> Utilice la función creada en el punto 1.1 para verificar sus resultados.

In [85]:
exploratory_data_analysis(df_grades)

##### 1. El DataFrame tiene 875 filas y 9 columnas

##### 2. Columnas del DataFrame:
['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

##### 3. Primeras 5 filas:

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
1  standard               completed        80.0           78.0           81.0  
2  standard                    none        73.0           66.0           63.0  
3  standard               completed        81.0           75.0           76.0  
4  standard               completed        94.0           86.0           87.0

##### Últimas 5 filas:

names  gender race/ethnicity parental level of education  \
870   Richard Young    male        group D                 high school   
871   Wanda Russell  female        group B                 high school   
872  Marina Zeigler  female        group C           bachelor's degree   
873   Laurie Carter  female        group B            some high school   
874    Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
870      standard                    none        69.0           75.0   
871  free/reduced               completed        23.0           44.0   
872  free/reduced               completed        66.0           74.0   
873      standard               completed        54.0           61.0   
874      standard               completed        68.0           80.0   

     writing score  
870           71.0  
871           36.0  
872           81.0  
873           62.0  
874           76.0

##### Muestreo aleatorio de 5 filas:

names  gender race/ethnicity parental level of education  \
208        Edward Lee    male        group D            some high school   
469  Anthony Anderson    male        group D                 high school   
336    David Martinez    male        group E          associate's degree   
269         Mary Levy  female        group D                some college   
2         Brian Young    male        group C            some high school   

            lunch test preparation course  math score  reading score  \
208      standard               completed        89.0           88.0   
469      standard                    none        88.0           78.0   
336  free/reduced               completed       100.0          100.0   
269  free/reduced                    none        60.0           66.0   
2        standard                    none        73.0           66.0   

     writing score  
208           82.0  
469           75.0  
336           93.0  
269           70.0  
2             63.0

##### 4. Descripción numérica del DataFrame:

math score  reading score  writing score
count  875.000000     875.000000     875.000000
mean    66.225143      69.195429      68.114286
std     15.076716      14.629270      15.174220
min      0.000000      17.000000      10.000000
25%     57.000000      59.000000      57.500000
50%     66.000000      70.000000      69.000000
75%     76.500000      79.000000      79.000000
max    100.000000     100.000000     100.000000

##### 5. Cantidad de valores nulos por columna:

names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

##### 6. Cantidad de valores únicos por columna:

names                          875
gender                           2
race/ethnicity                   5
parental level of education      6
lunch                            2
test preparation course          2
math score                      79
reading score                   72
writing score                   77
dtype: int64

##### 7. Total de filas con el nombre duplicado: 0

##### 8. Filas duplicadas por el nombre:

Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [86]:
def nota_chilena(dataframe_in, columna):
  dataframe = dataframe_in.copy()
  return (dataframe.loc[:, columna] * 0.06 + 1).round(2)

In [87]:
df_grades['math score'] = nota_chilena(df_grades, "math score")
df_grades['reading score'] = nota_chilena(df_grades, "reading score")
df_grades['writing score'] = nota_chilena(df_grades, "writing score")

df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
870           5.26  
871           3.16  
872           5.86  
873           4.72  
874           5.56  

[875 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [89]:
def alumnos_promedio(dataframe_in):
  dataframe = dataframe_in.copy()
  dataframe['GPA'] = dataframe.iloc[:,6:].mean(axis=1).round(2)
  return dataframe

In [90]:
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1  Charles Linstrom    male        group A           bachelor's degree   
2       Brian Young    male        group C            some high school   
3    Howard Jimenez    male        group E            some high school   
4      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
1  standard               completed        5.80           5.68           5.86   
2  standard                    none        5.38           4.96           4.78   
3  standard               completed        5.86           5.50           5.56   
4  standard               completed        6.64           6.16           6.22   

    GPA  
0  3.58  
1  5.78  
2  5.04  
3  5.64  
4  6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [94]:
def reprobados_destacados(dataframe_in):
    dataframe = dataframe_in.copy()
    # alumnos reprobados
    df_reprobados = dataframe[dataframe["GPA"] < 4.0].reset_index(drop=True)
    # reemplazo R
    dataframe["GPA"] = dataframe_in["GPA"].apply(lambda x: "R" if x < 4 else x)
    # alumnos detacados
    df_destacados = dataframe[
        (df_grades["math score"] > 5.8)
        & (df_grades["reading score"] > 5.8)
        & (df_grades["writing score"] > 5.8)
    ].reset_index(drop=True)
    # ordenamos por GPA
    df_destacados = df_destacados.sort_values(
        ascending=False, by="GPA"
    ).reset_index(drop=True)
    
    return df_reprobados, dataframe, df_destacados

In [95]:
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)

In [96]:
dataframe1

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1       Mikki Amaya  female        group A                 high school   
2     Samuel Dougal    male        group C             master's degree   
3   Rodger Talavera    male        group B                 high school   
4    Morris Johnson    male        group A          associate's degree   
..              ...     ...            ...                         ...   
81   Charles Harian    male        group C            some high school   
82    Wendy Gunnell  female        group D            some high school   
83      Diana Costa  female        group C                 high school   
84      Amy Sieving  female        group C                 high school   
85    Wanda Russell  female        group B                 high school   

           lunch test preparation course  math score  reading score  \
0       standard                    none        3.22           3.76   
1   free/reduced               completed        3.04           3.88   
2   free/reduced               completed        3.76           3.52   
3   free/reduced                    none        3.16           2.74   
4   free/reduced               completed        3.40           4.30   
..           ...                     ...         ...            ...   
81  free/reduced               completed        3.70           4.12   
82  free/reduced                    none        2.62           3.04   
83  free/reduced                    none        3.46           3.76   
84      standard                    none        2.74           2.74   
85  free/reduced               completed        2.38           3.64   

    writing score   GPA  
0            3.76  3.58  
1            3.46  3.46  
2            3.76  3.68  
3            2.62  2.84  
4            4.18  3.96  
..            ...   ...  
81           3.94  3.92  
82           2.92  2.86  
83           3.58  3.60  
84           2.80  2.76  
85           3.16  3.06  

[86 rows x 10 columns]

In [97]:
dataframe2

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score   GPA  
0             3.76     R  
1             5.86  5.78  
2             4.78  5.04  
3             5.56  5.64  
4             6.22  6.34  
..             ...   ...  
870           5.26   5.3  
871           3.16     R  
872           5.86  5.42  
873           4.72  4.54  
874           5.56  5.48  

[875 rows x 10 columns]

In [98]:
dataframe3

names  gender race/ethnicity parental level of education  \
0       Richard Pryor    male        group E           bachelor's degree   
1       Sandra Pompey  female        group E          associate's degree   
2   Elizabeth Beasley  female        group E           bachelor's degree   
3           Emma Gray  female        group E           bachelor's degree   
4        Emma Brasher  female        group D            some high school   
..                ...     ...            ...                         ...   
89     Rolando Widmer    male        group C                 high school   
90  Lucille Stanphill  female        group C                 high school   
91       Steven Kelly    male        group D                some college   
92        Peter Perez    male        group D             master's degree   
93        Harry Berry    male        group B          associate's degree   

       lunch test preparation course  math score  reading score  \
0   standard               completed        7.00           7.00   
1   standard                    none        7.00           7.00   
2   standard                    none        7.00           7.00   
3   standard               completed        6.94           7.00   
4   standard               completed        6.82           7.00   
..       ...                     ...         ...            ...   
89  standard               completed        5.92           6.04   
90  standard                    none        5.86           6.04   
91  standard                    none        5.86           5.92   
92  standard                    none        5.86           5.86   
93  standard               completed        5.86           5.92   

    writing score   GPA  
0            7.00   7.0  
1            7.00   7.0  
2            7.00   7.0  
3            7.00  6.98  
4            7.00  6.94  
..            ...   ...  
89           5.92  5.96  
90           5.92  5.94  
91           6.04  5.94  
92           6.04  5.92  
93           5.92   5.9  

[94 rows x 10 columns]

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [100]:
"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

# Código para quintiles
notas = pd.qcut(
    df_grades["GPA"], q=5, labels=["Bajísimo", "Bajo", "Medio", "Alto", "Muy alto"]
)
df_grades["GPA cuantiles"] = notas

fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig



In [ ]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig


A mayor nivel de estudio de los padres, o sea mejor degree, los GPA tienden a ser mejores, se ve como quienes tienen master tienen el más alto y de school es más bajo y se desplaza horizontalmente a la derecha .

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

In [102]:
df_grades_1 = pd.read_json('data/students_grades_1.json')
df_grades_2 = pd.read_json('data/students_grades_2.json')

df_grades = pd.concat([df_grades_1, df_grades_2], axis=0)
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos] **


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


In [103]:
df_grades["GPA"] = df_grades.mean(axis=1, numeric_only=True)

#simplificacion mapeo
df_grades["simple parental level of education"] = df_grades[
    "parental level of education"
].map(
    {
        "some high school": "school",
        "some college": "school",
        "high school": "school",
        "bachelor's degree": "college",
        "associate's degree": "college",
        "master's degree": "postgraduate",
    }
)

# agregacion para promedio de notas
agg = (
    df_grades.groupby(["race/ethnicity", "simple parental level of education"])
    .mean(numeric_only=True)
    .round(2)
)

# agregacion para obtener conteo alumnos por grupo
count = (
    df_grades.groupby(["race/ethnicity", "simple parental level of education"])
    .count()
    .round(2)
)

agg["count"] = count["GPA"]

percentage = (count["GPA"] / len(df_grades) * 100).round(2).astype(str) + " %"

agg["percentage"] = percentage
agg

math score  reading score  \
race/ethnicity simple parental level of education                              
group A        college                                   4.74           5.00   
               postgraduate                              4.69           5.23   
               school                                    4.57           4.73   
group B        college                                   5.07           5.26   
               postgraduate                              4.91           5.69   
               school                                    4.69           4.89   
group C        college                                   5.02           5.37   
               postgraduate                              4.92           5.14   
               school                                    4.76           5.02   
group D        college                                   5.11           5.25   
               postgraduate                              5.22           5.54   
               school                                    5.02           5.13   
group E        college                                   5.54           5.45   
               postgraduate                              5.54           6.03   
               school                                    5.40           5.31   

                                                   writing score   GPA  count  \
race/ethnicity simple parental level of education                               
group A        college                                      4.89  4.88     24   
               postgraduate                                 5.35  5.09      2   
               school                                       4.56  4.62     51   
group B        college                                      5.19  5.18     54   
               postgraduate                                 5.55  5.38      5   
               school                                       4.76  4.78    107   
group C        college                                      5.35  5.25    102   
               postgraduate                                 5.10  5.06     15   
               school                                       4.92  4.90    155   
group D        college                                      5.25  5.20     70   
               postgraduate                                 5.73  5.50     20   
               school                                       5.11  5.09    149   
group E        college                                      5.45  5.48     52   
               postgraduate                                 5.89  5.82      6   
               school                                       5.16  5.29     63   

                                                  percentage  
race/ethnicity simple parental level of education             
group A        college                                2.74 %  
               postgraduate                           0.23 %  
               school                                 5.83 %  
group B        college                                6.17 %  
               postgraduate                           0.57 %  
               school                                12.23 %  
group C        college                               11.66 %  
               postgraduate                           1.71 %  
               school                                17.71 %  
group D        college                                 8.0 %  
               postgraduate                           2.29 %  
               school                                17.03 %  
group E        college                                5.94 %  
               postgraduate                           0.69 %  
               school                                  7.2 %

In [104]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

**Resultado Esperado de la tabla**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [106]:
other_grades = pd.read_csv('data/other_grades.csv', sep=',')
df_grades_outer = df_grades.merge(other_grades, on='names', how='outer')
df_grades_left = df_grades.merge(other_grades, on='names', how='left')
df_grades_right = df_grades.merge(other_grades, on='names', how='right')
df_grades_inner = df_grades.merge(other_grades, on='names', how='inner')

In [107]:
df_grades_outer.shape

(1000, 13)

In [108]:
df_grades_left.shape

(875, 13)

In [109]:
df_grades_right.shape

(1000, 13)

In [110]:
df_grades_inner.shape

(875, 13)

Con el outer tendré todos los registros, con el left los registros solamente que están en los datos de la izquierda y con el right, como nos entrega 1000, similar al outer, es porque están todos los datos, y el inner es donde cruzan todos, y como vimos en el left el minimo es 875 que es cuando todos hacen match para la columna dada.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>